include/interpolator.hpp

```diff
@@ -0,0 +1,45 @@
+#ifndef __INTERPOLATOR_H__
+#define __INTERPOLATOR_H__
+#include <vector>
+
+#include <neso_particles.hpp>
+
+namespace NP = NESO::Particles;
+namespace NESO {
+/**
+ *  Base class that defines functions and variables needed to read interpolate
+ * data
+ *
+ * Derived classes must override interpolate() in order to populate y_output (1d
+ * interpolator)
+ */
+
+class Interpolator {
+public:
+  Interpolator(std::vector<double> x_data, std::vector<double> y_data,
+               NP::SYCLTargetSharedPtr sycl_target)
+      : m_sycl_target(sycl_target), m_x_data(x_data), m_y_data(y_data) {
+    // calculate change in y from between each vector array position
```

Comment:
As the loop of `push_back` is of known size you add `dy.reserve(m_y_data.size())` to explicitly avoid reallocations.

Reply:
Done